content/en/blog/_posts/2024-05-01-cri-streaming-explained.md

```diff
@@ -0,0 +1,341 @@
+
+All RPCs of the CRI either use the [gRPC unary calls](https://grpc.io/docs/what-is-grpc/core-concepts/#unary-rpc)
+for communication or the [server side streaming](https://grpc.io/docs/what-is-grpc/core-concepts/#server-streaming-rpc)
+feature (only `GetContainerEvents` right now). This means that mainly all RPCs
+retrieve a single client request and have to return a single server response.
+The same applies to **Exec**, **Attach** and **PortForward**, where [their protocol definition](https://github.com/kubernetes/cri-api/blob/63929b3/pkg/apis/runtime/v1/api.proto#L94-L99)
+looks like this:
+
+```protobuf
+// Exec prepares a streaming endpoint to execute a command in the container.
+rpc Exec(ExecRequest) returns (ExecResponse) {}
+```
+
+```protobuf
+// Attach prepares a streaming endpoint to attach to a running container.
+rpc Attach(AttachRequest) returns (AttachResponse) {}
+```
+
+```protobuf
+// PortForward prepares a streaming endpoint to forward ports from a PodSandbox.
+rpc PortForward(PortForwardRequest) returns (PortForwardResponse) {}
+```
+
+The requests carry everything required to allow the server to do the work,
+for example the `ContainerId` or command (`Cmd`) to be run in case of `Exec`.
+More interestingly, all of their responses only contain a `url`:
+
+```protobuf
+message ExecResponse {
+    // Fully qualified URL of the exec streaming server.
+    string url = 1;
+}
+```
+
+```protobuf
+message AttachResponse {
+    // Fully qualified URL of the attach streaming server.
+    string url = 1;
+}
+```
+
+```protobuf
+message PortForwardResponse {
+    // Fully qualified URL of the port-forward streaming server.
+    string url = 1;
+}
+```
+
+Why is it implemented like that? Well, [the original design document](https://docs.google.com/document/d/1MreuHzNvkBW6q7o_zehm1CBOBof3shbtMTGtUpjpRmY)
+for those RPCs even predates [Kubernetes Enhancements Proposals (KEPs)](https://github.com/kubernetes/enhancements)
+and was originally outlined back in 2016. The kubelet had a native
+implementation for **Exec**, **Attach** and **PortForward** before the
+initiative of bringing the functionality to the CRI had started. Before that,
+everything was bound to [Docker](https://www.docker.com) or the later abandoned
+container runtime [rkt](https://github.com/rkt/rkt).
+
+The CRI related design document also elaborates on the option to use native RPC
+streaming for exec, attach and port forward. The downsides outweighed this
+approach: the kubelet would still create a network bottleneck and future
+runtimes would not be free in choosing the server implementation details. Also,
+another option that the Kubelet implements a portable, runtime-agnostic solution
+has been abandoned over the final one, because this would mean another project
+to maintain which nevertheless would be runtime dependent.
+
+This means, that the basic flow for **Exec**, **Attach** and **PortForward**
+was proposed to look like this:
+
+<!-- prettier-ignore -->
```

Comment:
Is it necessary to keep this comment, or can we remove it?
```suggestion
```

Reply:
Removed them.